### PR TITLE
feat: remember stopped extensions across restart

### PR DESCRIPTION
**Pull request by deboer-tim**
### What does this PR do?

Introduces a new hidden configuration value 'extensions.disabled' to store an array of disabled extension ids. Disabled extensions are loaded but not activated on startup (with a log message).

When extensions are manually deactivated they're added to the list. Since shutdown and reload also deactivate extensions, a system argument is used to differentiate between the two.

When extensions are (manually) started they are removed from the list before trying to load.

Also renamed *activation > *activating so that the log is consistent.

### Screenshot/screencast of this PR

N/A.

### What issues does this PR fix or reference?

Fixes issue #268.

### How to test this PR?

Go to Settings > Extensions and stop an extension. Restart to confirm extension is still stopped. Start it; stop one or two others; restart and check again. No new errors in the log.

### CODE REVIEW

**Review by jeffmaury — COMMENTED** [review]
I had OpenShift Local extension installed and it says default extension, cannot be removed
Not related to this PR as it is also on main

File: packages/main/src/plugin/extension-loader.ts [hunk 1]
```diff
@@ -212,6 +214,22 @@ export class ExtensionLoader {
...
   async init(): Promise<void> {
+    const extensionConfiguration: IConfigurationNode = {
+      id: 'preferences.extensions',
+      title: 'Extensions',
+      type: 'object',
+      properties: {
+        [ExtensionSettings.SectionName + '.' + ExtensionSettings.Disabled]: {
+          description: 'Disabled extensions',
+          type: 'array',
+          default: undefined,
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:225]: should default be [] instead of undefined ?

File: packages/main/src/plugin/extension-loader.ts [hunk 2]
```diff
@@ -1036,7 +1075,7 @@ export class ExtensionLoader {
     }
   }
 
-  async deactivateExtension(extensionId: string): Promise<void> {
+  async deactivateExtension(extensionId: string, system?: boolean): Promise<void> {
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts]: quickly looking at the `system` parameter I don't understand what it will do. Like is it because it's a system extension ?

maybe rename it to `shutdown` or `systemShutdown` or `systemOrder` ?

or just what is triggering the call: `stopAllExtensions`

File: packages/main/src/plugin/extension-loader.ts [hunk 3]
```diff
@@ -1047,11 +1086,26 @@ export class ExtensionLoader {
     this.extensionState.set(extension.id, 'stopping');
     this.apiSender.send('extension-stopping');
 
+    if (!system) {
+      // extension was explicitly disabled, save the configuration
+      let disabledExtensions = this.getDisabledExtensionIds();
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts]: could help if the method return always array (including empty array) and not undefined

File: packages/main/src/plugin/extension-loader.ts [hunk 4]
```diff
@@ -1047,11 +1086,26 @@ export class ExtensionLoader {
...
+
+      if (!disabledExtensions?.includes(extensionId)) {
+        if (disabledExtensions) {
+          disabledExtensions.push(extensionId);
+        } else {
+          disabledExtensions = [extensionId];
+        }
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts]: maybe we should use a Set instead of Array if you only want at more one element

**Review by benoitf — COMMENTED** [review]
to me, a disabled extension is an extension that is not started, not an extension that is not activated.

For example if extension contribute menus, etc. I don't want to see these items. (because it's not only the call to activate function)

I want 0 code of this extension being shown in Podman Desktop

**Comment by deboer-tim** [comment]
> I had OpenShift Local extension installed and it says default extension, cannot be removed
> Not related to this PR as it is also on main

I had seen this yesterday, but ran out of time to investigate and confirm it wasn't just my workspace. Confirmed they switch after first restart, opened issue 3149 and will look into it separately.

File: packages/main/src/plugin/extension-loader.ts (hunk 1 above)
Comment (deboer-tim) [reply packages/main/src/plugin/extension-loader.ts:225]: Sure, that would remove the need for some if/?s as well.

File: packages/main/src/plugin/extension-loader.ts (hunk 2 above)
Comment (deboer-tim) [reply packages/main/src/plugin/extension-loader.ts]: It is also called when reloading (line 519 above), so I was trying to differentiate between 'the system' starting and stopping extensions vs user action. Better name? Or I could refactor deactivate so that it's like startExtension (method is only called by users, the system calls analyze/load directly).

File: packages/main/src/plugin/extension-loader.ts (hunk 3 above)
Comment (deboer-tim) [reply packages/main/src/plugin/extension-loader.ts]: Agreed.

File: packages/main/src/plugin/extension-loader.ts (hunk 4 above)
Comment (deboer-tim) [reply packages/main/src/plugin/extension-loader.ts]: Ordering doesn't matter either, so yes. I'll take a look at the config registry separately and see if it makes sense to add support for Sets.

**Comment by deboer-tim** [comment]
> to me, a disabled extension is an extension that is not started, not an extension that is not activated.
> 
> For example if extension contribute menus, etc. I don't want to see these items. (because it's not only the call to activate function)
> 
> I want 0 code of this extension being shown in Podman Desktop

100% agreed. I was not taking a close enough look at what is happening in loadExtension() and it will need to be refactored. Currently it is adding contributions before registering that the extension exists, then trying to activate it. It should register first, check if disabled, then add contributions and start. I will refactor.

**Comment by benoitf** [comment]
@deboer-tim I think you can apply your filter here https://github.com/containers/podman-desktop/blob/main/packages/main/src/plugin/extension-loader.ts#L343-L354

you remove disabled extensions from the extensions array to be loaded

**Comment by deboer-tim** [comment]
I tried that first - doesn't work b/c the extension isn't added to analyzedExtensions until later (https://github.com/containers/podman-desktop/blob/7ad645558d8c9d5bedbc568200068fd569399b6f/packages/main/src/plugin/extension-loader.ts#L533), so it doesn't appear on the extensions page and you can never start it again.

That's the refactoring I think I need to do: if the extension is valid register it as a (stopped) analyzed extension first, before deciding whether to start/adding contributions/etc.

**Comment by benoitf** [comment]
ah yes correct, need to add a 'disabled' field to the AnalyzedExtension 👍

**Comment by deboer-tim** [comment]
Updated branch pushed, ready for new review.

Refactored previous commit to have clean extension states:
- Adding an extension to analyzedExtensions is 'registering it with the system'. On startup we load and register all known extensions, then trigger activation for all that haven't been disabled.
- activate/deactivateExtension() are how extensions get started & stopped. Activate registers the contributions (refactored out of loadExtension to new method) and calls it's main; deactivate does the reverse.
- start/stopExtension are how the UI/user requests extensions to be activated or deactivated. stopExtension is new and allows us to differentiate between the system deactivating an extension (e.g. during shutdown or reload) vs the user.

With this refactoring it is much cleaner in that everything to do with loading and unloading an extension is in activate/deactivate (the 'implementation' of starting and stopping), and all external requests come through start/stop (the same as what was already exposed in the API). I've tested all the paths I can think of and no longer see contributions for disabled/stopped extensions.

Additional notes:
- The configuration registry does not support sets. I spent a while trying to add it but haven't been successful yet since you have to write replacer/reviver functions for JSON.stringify/parse. We can either live with an array for now or block this until it supports sets.
- Likewise, if you use `default: []` the configuration registry won't save the property at all. I've narrowed this down and will open a bug, but this PR uses `| []` on line 275 instead.
- Changed the test from a failing activation function to a missing function. This wasn't caught before and would cause no extensions to load, so I extended the try block to catch this.

**Review by benoitf — COMMENTED** [review]
it looks like it is changing the behavior of a lifecycle of an extension
Like contributions being part of the 'activate' workflow while the contributions should be added when we load the extension. An extension can be loaded but not activated ( now it's auto-activated) but activation is independent of the load of an extension where we load everything that is contributed

maybe I miss something in a quick look at the code ?

**Comment by deboer-tim** [comment]
> it looks like it is changing the behavior of a lifecycle of an extension Like contributions being part of the 'activate' workflow while the contributions should be added when we load the extension. An extension can be loaded but not activated ( now it's auto-activated) but activation is independent of the load of an extension where we load everything that is contributed
> 
> maybe I miss something in a quick look at the code ?

No, I think you got that right. The biggest change is moving registering contributions to activate(). Maybe we need some renaming, but now:
- load = register extension into the list, check for missing dependencies, do prep work *but don't actually start anything*
- activate = add contributions and activate extension

This cleanly separates loading extension metadata and adding it into the list from triggering any extension code/UI changes. It's essentially adding a state where (disabled) extensions can be in analyzedExtensions but have no extensionState, which is still reported as stopped in the UI.

**Comment by deboer-tim** [comment]
@benoitf Could you please give another review?

File: packages/main/src/plugin/extension-loader.ts [hunk 5]
```diff
@@ -499,6 +535,10 @@ export class ExtensionLoader {
   }
 
   async loadExtension(extension: AnalyzedExtension, checkForMissingDependencies = false): Promise<void> {
+    this.analyzedExtensions.set(extension.id, extension);
+    this.extensionState.delete(extension.id);
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:539]: shouldn't be state being 'loading' instead of deleting the state ? (it seems you only moved a line but it feels strange that when loading we delete the state)

File: packages/main/src/plugin/extension-loader.ts [hunk 6]
```diff
@@ -499,6 +535,10 @@ export class ExtensionLoader {
   }
 
   async loadExtension(extension: AnalyzedExtension, checkForMissingDependencies = false): Promise<void> {
+    this.analyzedExtensions.set(extension.id, extension);
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:538]: it feels strange that we set objects as being 'analyzed' at the beginning of the method

I would expect analyzeMethod set analyzedExtensions objects and probably loadExtension should store to loadedExtensions map but only at the end of the method ?

File: packages/main/src/plugin/extension-loader.ts [hunk 7]
```diff
@@ -562,9 +567,12 @@ export class ExtensionLoader {
         this.watcherExtensions.set(extension.id, extensionWatcher);
       }
 
-      const runtime = this.loadRuntime(extension);
-
-      await this.activateExtension(extension, runtime);
+      // activate the extension if it hasn't been disabled
+      if (!this.getDisabledExtensionIds().includes(extension.id)) {
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:571]: it doesn't feel natural to store disabled state in a separate map/array

we already map of extensions, ids, etc.

Probably the state should go to the extension object directly. It would avoid another indirection level and keep consistency between extensions and the disabled extensions.

I would avoid usage  of `double negation` like instead of not disabled ((!this.getDisabledExtensionIds), check enabled and if state of enablement is in the extension it's easier.

like `if (extension.isEnabled() {
   await this.activateExtension()
`

File: packages/main/src/plugin/extension-loader.ts [hunk 8]
```diff
@@ -562,9 +567,12 @@ export class ExtensionLoader {
...
+        await this.activateExtension(extension);
+      } else {
+        console.log('Not activating disabled extension (' + extension.id + ')');
+      }
     } catch (err) {
       this.extensionState.set(extension.id, 'failed');
       this.extensionStateErrors.set(extension.id, err);
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:578]: state should also go to the extension object

File: packages/main/src/plugin/extension-loader.ts [hunk 9]
```diff
@@ -1091,6 +1132,18 @@ export class ExtensionLoader {
...
 
+  async stopExtension(extensionId: string): Promise<void> {
+    // the user explicitly disabled the extension, save the configuration
+    const disabledExtensions = this.getDisabledExtensionIds();
+
+    if (!disabledExtensions.includes(extensionId)) {
+      disabledExtensions.push(extensionId);
+      this.setDisabledExtensionIds(disabledExtensions);
+    }
+
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:1143]: here it would be easier if disabled is already stored in the extension.

it's strange also that we don't check if extension exists before adding it to the list ?

Comment (benoitf) [reply packages/main/src/plugin/extension-loader.ts:1143]: I'm not sure to understand also what it does.

if I stop and extension and the extension is not in the disabled list, it adds this extension as being disabled

so it is not a stop but a disableExtension method

File: packages/main/src/plugin/extension-loader.ts [hunk 10]
```diff
@@ -1100,6 +1153,16 @@ export class ExtensionLoader {
...
+      const disabledExtensionIds = this.getDisabledExtensionIds();
+      const index = disabledExtensionIds.indexOf(extensionId);
+      if (index > -1) {
+        // extension was disabled but the user has activated it,
+        // clean the configuration up
+        disabledExtensionIds.splice(index, 1);
+
+        this.setDisabledExtensionIds(disabledExtensionIds);
+      }
+
```
Comment (benoitf) [inline packages/main/src/plugin/extension-loader.ts:1165]: if state is in the extension we do only
`if (extension.isEnabled()...)`

File: packages/main/src/plugin/index.ts [hunk 11]
```diff
@@ -1349,9 +1349,9 @@ export class PluginSystem {
     });
 
     this.ipcHandle(
-      'extension-loader:deactivateExtension',
+      'extension-loader:stopExtension',
       async (_listener: Electron.IpcMainInvokeEvent, extensionId: string): Promise<void> => {
-        return this.extensionLoader.deactivateExtension(extensionId);
+        return this.extensionLoader.stopExtension(extensionId);
```
Comment (benoitf) [inline packages/main/src/plugin/index.ts:1354]: I guess it's disableExtension now ?

File: packages/preload/src/index.ts [hunk 12]
```diff
@@ -942,7 +942,7 @@ function initExposure(): void {
   });
 
   contextBridge.exposeInMainWorld('stopExtension', async (extensionId: string): Promise<void> => {
-    return ipcInvoke('extension-loader:deactivateExtension', extensionId);
+    return ipcInvoke('extension-loader:stopExtension', extensionId);
```
Comment (benoitf) [inline packages/preload/src/index.ts:945]: disableExtension

**Review by benoitf — COMMENTED** [review]
I think state was added later (as a separate map), then stateError was added (also as a separate map) and now you're trying to add enable/disable as a separate array.

I think as you're switching to always having extensions (to get state/errors/etc), everything needs to be merged in a single object.

for example when we list extensions we do
```
async listExtensions(): Promise<ExtensionInfo[]> {
    return Array.from(this.analyzedExtensions.values()).map(extension => ({
      state: this.extensionState.get(extension.id) || 'stopped',
      error: this.mapError(this.extensionStateErrors.get(extension.id)),

```

it should not use for state and error another data, it should already be there, in the maintained Map


so in short, one map with extensionId as a key

then state, enabled/disabled, errors, manifest, etc it's all in the same object.

Else it comes to be a nightmare as we need to maintain state across 4 different maps/arrays in the same Class for the same stuff

**Comment by cdrage** [comment]
Just going through some old PR's, @deboer-tim what's the status of the PR?
